d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -260,22 +312,105 @@ void handleD2URIMapResponse(DiscoveryResponseData data)
     handleResourceRemoval(data.getRemovedResources(), data.getResourceType());
   }
 
+  /**
+   * Handles glob collection responses by looking up the existing {@link ResourceSubscriber}'s data and applying the
+   * patch received from the xDS server.
+   */
+  private void handleD2URICollectionResponse(DiscoveryResponseData data)
+  {
+    Map<String, D2URIMapUpdate> updates = new HashMap<>();
+    List<String> errors = new ArrayList<>();
+
+    Set<String> removedClusters = new HashSet<>();
+
+    data.foreach((resourceName, resource) ->
+    {
+      D2UriIdentifier uriId = D2UriIdentifier.parse(resourceName);
+      if (uriId == null)
+      {
+        String msg = String.format("Ignoring D2URI resource update with invalid name: %s", resourceName);
+        _log.warn(msg);
+        errors.add(msg);
+        return;
+      }
+
+      ResourceSubscriber subscriber =
+          getResourceSubscriberMap(ResourceType.D2_URI_MAP).get(uriId.getClusterResourceName());
+      if (subscriber == null)
+      {
+        String msg =
+            String.format("Ignoring D2URI resource update for untracked cluster: %s", uriId.getClusterResourceName());
+        _log.warn(msg);
+        errors.add(msg);
+        return;
+      }
+
+      // Get or create a new D2URIMapUpdate which is a copy of the existing data for that cluster.
+      D2URIMapUpdate update = updates.computeIfAbsent(uriId.getClusterResourceName(), k ->
+      {
+        D2URIMapUpdate currentData = (D2URIMapUpdate) subscriber._data;
+        if (currentData == null || currentData.isEmpty())
+        {
+          return new D2URIMapUpdate(new HashMap<>());
+        }
+        else
+        {
+          return new D2URIMapUpdate(new HashMap<>(currentData.getURIMap()));
+        }
+      });
+
+      // If the resource is null, it's being deleted
+      if (resource == null)
+      {
+        // This is the special case where the entire collection is being deleted. This either means the client
+        // subscribed to a cluster that does not exist, or all hosts stopped announcing to the cluster.
+        if ("*".equals(uriId.getUriName()))
+        {
+          removedClusters.add(uriId.getClusterResourceName());
+        }
+        else
+        {
+          // Else it's a standard delete for that host.
+          update.getURIMap().remove(uriId.getUriName());
+        }
+      }
+      else
+      {
+        try
+        {
+          XdsD2.D2URI uri = resource.getResource().unpack(XdsD2.D2URI.class);
+          update.getURIMap().put(uriId.getUriName(), uri);
+        }
+        catch (InvalidProtocolBufferException e)
+        {
+          _log.warn("Failed to unpack D2URI glob collection response", e);
+          errors.add("Failed to unpack D2URI glob collection response");
+        }
+      }
+    });
+    sendAckOrNack(data.getResourceType(), data.getNonce(), errors);
+
+    handleResourceUpdate(updates, ResourceType.D2_URI_MAP);
```

Comment:
if any of the update inside updates has an empty map, we won't be logging about it. Basically the log at line 313 is missing with glob collection enabled.

Reply:
synced offline. Actually empty maps should be valid for cluster uris maps since the uris under a cluster could have all de-announced legitimately. (as opposed to removing the UriMap resource entirely should be ignored).